### PR TITLE
[athena_2]: complete the task_2

### PR DESCRIPTION
**Pull request by martinx**
### What?

* brief: task_2 完成
* author: killing11l, badmicroctrl, martinx

### What changes are included in this PR?

* 完成stl_vector代码学习
* 整理并编写学习笔记
* 编写测试代码

### Are these changes tested?

* 已测试

### CODE REVIEW

File: stl/athena/killing11l/src/task_2/thread_pool.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    using ret_type = typename std::result_of<FUN(ARGS...)>::type;
+    auto task = std::make_shared<std::packaged_task<ret_type()>>(
+        std::bind(std::forward<FUN>(f), std::forward<ARGS>(args)...));
+
+    auto res = task->get_future();
+    {
+      std::unique_lock<std::mutex> lock(mtx_);
+      tasks_.emplace([task] { (*task)(); });
+    }
+    cond_.notify_all();
```
Comment (Light-City) [inline stl/athena/killing11l/src/task_2/thread_pool.h:29]: may be use notify_one

File: stl/athena/martinx/src/task_2/observer.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+int Subject::getState() { return state_; }
+
+void Subject::notifyObservers() {
+  for (auto it = observers_.begin(); it != observers_.end(); ++it) {
+    (*it)->notify();
+  }
+}
+
+Observer::Observer(Subject* subject) : subject_(*subject) {
```
Comment (Light-City) [inline stl/athena/martinx/src/task_2/observer.cpp:31]: consider std::move(subject)

File: stl/athena/martinx/src/task_2/observer.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  void registerObserver(Observer* o);
+  void removeObserver(Observer* o);
+  void notifyObservers();
+
+  //  演示用
+  void setState(int state);
+  int getState();
+
+ protected:
+  std::vector<Observer*> observers_;
```
Comment (Light-City) [inline stl/athena/martinx/src/task_2/observer.h:33]: don't forget to release

**Comment by martinx** [comment]
Thanks, we fixed the above issues. Please help us to review it again. ❤️

File: stl/athena/killing11l/src/task_2/thread_pool.h (hunk 1 above)
Comment (Killing11L) [reply stl/athena/killing11l/src/task_2/thread_pool.h]: Ok, so notify_one is really better for this scenario.

File: stl/athena/martinx/src/task_2/observer.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
+#ifndef OBSERVER_H
+#define OBSERVER_H
+#include <vector>
+
+class Subject;
+
+class Observer : public std::enable_shared_from_this<Observer> {
```
Comment (Light-City) [inline stl/athena/martinx/src/task_2/observer.h:7]: why use enable_shared_from_this?

File: stl/athena/martinx/src/task_2/observer.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ public:
+  Observer();
+  ~Observer();
+
+  void notify();
+  std::shared_ptr<Observer> shared_observer();
```
Comment (Light-City) [inline stl/athena/martinx/src/task_2/observer.h:13]: nowhere else to call

**Review by Light-City — COMMENTED** [review]
lgtm

**Pull request merged**